src/domain/tradingview/helper.ts

```diff
@@ -0,0 +1,119 @@
+    try {
+      prices = await getChainlinkChartPricesFromGraph(symbol, period);
+    } catch (ex2) {
+      // eslint-disable-next-line no-console
+      console.warn("getChainlinkChartPricesFromGraph failed", ex2);
+      prices = [];
+    }
+  }
+  return prices;
+}
+
+async function getCurrentPriceOfToken(chainId, symbol) {
+  try {
+    const indexPricesUrl = getServerUrl(chainId, "/prices");
+    const indexPrices = await fetch(indexPricesUrl).then((res) => res.json());
+    let symbolInfo = getTokenBySymbol(chainId, symbol);
+    if (symbolInfo.isNative) {
+      symbolInfo = getWrappedToken(chainId);
+    }
+    return indexPrices[symbolInfo.address];
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.error(err);
+  }
+}
+
+export async function getHistoryBars({ ticker, resolution, chainId, isStable, countBack }) {
+  const period = supportedResolutions[resolution];
+  const bars = isStable ? getStablePriceData(period, countBack) : await getTokenChartPrice(chainId, ticker, period);
+  return bars.map(formatTimeInBar);
+}
+
+async function getLastHistoryBar(ticker, resolution, chainId) {
+  const prices = await getLimitChartPricesFromStats(chainId, ticker, resolution);
+  return formatTimeInBar({ ...prices[prices.length - 1], ticker });
+}
+
+const getLastBarAfterInterval = (function () {
+  let startTime = 0;
+  let lastBar, lastTicker, lastPeriod;
+
+  return async function main(ticker, period, chainId) {
+    const currentTime = Date.now();
+    if (currentTime - startTime > 15000 || lastTicker !== ticker || lastPeriod !== period) {
+      lastBar = await getLastHistoryBar(ticker, period, chainId);
+      startTime = currentTime;
+      lastTicker = ticker;
+      lastPeriod = period;
+    }
+    return lastBar;
+  };
+})();
+
+export async function getLiveBar({ ticker, resolution, chainId }) {
+  const period = supportedResolutions[resolution];
+  if (!ticker || !period || !chainId) return;
+
+  const periodSeconds = CHART_PERIODS[period];
+  const currentCandleTime = (Math.floor(Date.now() / 1000 / periodSeconds) * periodSeconds + timezoneOffset) * 1000;
+
+  let lastBar = await getLastBarAfterInterval(ticker, period, chainId);
+
+  if (!lastBar) return;
+
+  const currentPrice = await getCurrentPriceOfToken(chainId, ticker);
+  const averagePriceValue = parseFloat(formatAmount(currentPrice, USD_DECIMALS, 4));
+  if (lastBar.time && currentCandleTime === lastBar.time && ticker === lastBar.ticker) {
+    return {
+      ...lastBar,
+      close: averagePriceValue,
+      high: Math.max(lastBar.open, lastBar.high, averagePriceValue),
+      low: Math.min(lastBar.open, lastBar.low, averagePriceValue),
+      ticker,
+    };
+  } else {
+    const newBar = {
+      time: currentCandleTime,
+      open: lastBar.close,
+      close: averagePriceValue,
+      high: Math.max(lastBar.close, averagePriceValue),
+      low: Math.min(lastBar.close, averagePriceValue),
+      ticker,
+    };
+    lastBar = newBar;
```

Comment:
is there a reason to override lastBar value? or `let lastBar ...` can be changed to `const lastBar ...`?

Reply:
But, why redefine lastBar here?
This is a clear function without a cycle, every time you call it, **lastBar** will be re-initialized, it seems `lastBar = newBar` shouldn't have any effect

Correct me if I'm wrong, but otherwise it looks like magic)

Reply:
Sorry I was mistaken. Earlier the structure was such that the lastBar was outside the function. So we needed to replace with the new values. I will fix.